library/spdm_responder_lib/libspdm_rsp_set_certificate.c

```diff
@@ -0,0 +1,134 @@
+ * @retval RETURN_DEVICE_ERROR          A device error occurs when communicates with the device.
+ * @retval RETURN_SECURITY_VIOLATION    Any verification fails.
+ **/
+libspdm_return_t libspdm_get_response_set_certificate(void *context, size_t request_size,
+                                                      const void *request,
+                                                      size_t *response_size,
+                                                      void *response)
+{
+    spdm_set_certificate_request_t *spdm_request;
+    spdm_set_certificate_response_t *spdm_response;
+
+    libspdm_context_t *spdm_context;
+    bool result;
+    uint8_t slot_id;
+    bool need_reset;
+    size_t root_cert_hash_size;
+
+    libspdm_session_info_t *session_info;
+    libspdm_session_state_t session_state;
+
+    spdm_context = context;
+    spdm_request = (void*)request;
+
+    if (spdm_request->header.spdm_version != libspdm_get_connection_version(spdm_context)) {
+        return libspdm_generate_error_response(spdm_context,
+                                               SPDM_ERROR_CODE_VERSION_MISMATCH, 0,
+                                               response_size, response);
+    }
+
+    if (spdm_context->response_state != LIBSPDM_RESPONSE_STATE_NORMAL) {
+        return libspdm_responder_handle_response_state(spdm_context,
+                                                       spdm_request->header.request_response_code,
+                                                       response_size, response);
+    }
+
+    if (request_size <= sizeof(spdm_set_certificate_request_t)) {
+        return libspdm_generate_error_response(spdm_context,
+                                               SPDM_ERROR_CODE_INVALID_REQUEST, 0,
+                                               response_size, response);
+    }
+
+    if (spdm_context->connection_info.connection_state <
+        LIBSPDM_CONNECTION_STATE_NEGOTIATED) {
+        return libspdm_generate_error_response(
+            spdm_context,
+            SPDM_ERROR_CODE_UNEXPECTED_REQUEST, 0,
+            response_size, response);
+    }
+
+    if (spdm_context->last_spdm_request_session_id_valid) {
+        session_info = libspdm_get_session_info_via_session_id(
+            spdm_context,
+            spdm_context->last_spdm_request_session_id);
+        if (session_info == NULL) {
+            return libspdm_generate_error_response(
+                spdm_context,
+                SPDM_ERROR_CODE_UNEXPECTED_REQUEST, 0,
+                response_size, response);
+        }
+        session_state = libspdm_secured_message_get_session_state(
+            session_info->secured_message_context);
+        if (session_state != LIBSPDM_SESSION_STATE_ESTABLISHED) {
+            return libspdm_generate_error_response(
+                spdm_context,
+                SPDM_ERROR_CODE_UNEXPECTED_REQUEST, 0,
+                response_size, response);
+        }
+    }
+

```

Comment:
Can include a check that `SET_CERTIFICATE.CertChain.Length == request_size - sizeof(spdm_message_header_t)`

Reply:
We cannot add this check, because the transport may add alignment padding for request_size.

I think we can only do `SET_CERTIFICATE.CertChain.Length <= request_size - sizeof(spdm_message_header_t)`

Reply:
Yes. I have added the follow check. And the `max_cert_chain_size` is `request_size - sizeof(spdm_message_header_t)`.

```
    if (max_cert_chain_size < cert_chain_size + sizeof(spdm_cert_chain_t) + root_cert_hash_size) {
        return false;
    }

```

Reply:
Why put the check in `libspdm_write_certificate_to_nvm`?

Reply:
Make sense.  I have added the check in `libspdm_get_response_set_certificate`. 
And I have changed the parameters of the call function. Thanks.
```
    /* set certificate to NV*/
    result = libspdm_write_certificate_to_nvm(slot_id, cert_chain,
                                              cert_chain_size,
                                              &need_reset);
```